node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/NodePatcher.java

```diff
@@ -271,36 +289,55 @@ private boolean asBoolean(Inspector field) {
         return Optional.of(field).filter(Inspector::valid).map(this::asBoolean);
     }
 
-    private class PatchedNodes {
-        private final Map<String, Node> nodes = new HashMap<>();
+    private class PatchedNodes implements Mutex {
+        private final Map<String, NodeMutex> nodes = new HashMap<>();
         private final String hostname;
         private boolean fetchedChildren;
 
-        private PatchedNodes(Node node) {
-            this.hostname = node.hostname();
+        private PatchedNodes(NodeMutex nodeMutex) {
+            this.hostname = nodeMutex.node().hostname();
+            nodes.put(hostname, nodeMutex);
+            fetchedChildren = !nodeMutex.node().type().isHost();
+        }
 
-            nodes.put(hostname, node);
-            fetchedChildren = !node.type().isHost();
+        public NodeMutex nodeMutex() {
+            return nodes.get(hostname);
         }
 
         public Node node() {
-            return nodes.get(hostname);
+            return nodeMutex().node();
         }
 
         public List<Node> children() {
             if (!fetchedChildren) {
-                memoizedNodes.get().childrenOf(hostname).forEach(this::update);
+                memoizedNodes.get()
+                        .childrenOf(hostname)
+                        .forEach(node -> nodeRepository.lockAndGet(node)
+                                .ifPresent(nodeMutex -> nodes.put(nodeMutex.node().hostname(), nodeMutex)));
                 fetchedChildren = true;
             }
-            return nodes.values().stream().filter(node -> !node.type().isHost()).collect(Collectors.toList());
+            return nodes.values().stream()
+                    .map(NodeMutex::node)
+                    .filter(node -> !node.type().isHost())
+                    .collect(Collectors.toList());
         }
 
         public void update(Node node) {
-            nodes.put(node.hostname(), node);
+            NodeMutex currentNodeMutex = nodes.get(node.hostname());
+            if (currentNodeMutex == null) {
+                throw new IllegalStateException("unable to update non-existing child: " + node.hostname());
+            }
+
+            nodes.put(node.hostname(), currentNodeMutex.with(node));
         }
 
         public List<Node> nodes() {
-            return List.copyOf(nodes.values());
+            return List.copyOf(nodes.values().stream().map(NodeMutex::node).collect(Collectors.toList()));
```

Comment:
Double copy, considering collecting to `Collectors.toUnmodifiableList()` instead.

Reply:
Removed List.copyOf().